### PR TITLE
Allow sending messages prior to closing the connection

### PR DESCRIPTION
**Pull request by mtrudel**
Related to https://github.com/phoenixframework/websock/issues/11

### CODE REVIEW

File: lib/websock.ex [hunk 1]
```diff
@@ -24,19 +24,26 @@ defmodule WebSock do
...
   @typedoc "The result as returned from init, handle_in, handle_control & handle_info calls"
   @type handle_result ::
-          {:push, {opcode(), message()} | [{opcode(), message()}], state()}
-          | {:reply, term(), {opcode(), message()} | [{opcode(), message()}], state()}
+          {:push, messages(), state()}
+          | {:reply, term(), messages(), state()}
           | {:ok, state()}
           | {:stop, {:shutdown, :restart} | term(), state()}
           | {:stop, term(), close_detail(), state()}
+          | {:stop, term(), close_detail(), messages(), state()}
```
Comment (mtrudel) [inline lib/websock.ex:37]: This is the salient change in this PR; the rest of the noise here is just to clean up typespecs

File: lib/websock.ex [hunk 2]
```diff
@@ -13,7 +13,7 @@ defmodule WebSock do
   @type state :: term()
 
   @typedoc "The structure of a sent or received WebSocket message body"
-  @type message :: iodata() | nil
+  @type payload :: iodata() | nil
```
Comment (mtrudel) [inline lib/websock.ex]: I've verified that no direct dependents of WebSock make use of this typespec

File: lib/websock.ex [hunk 3]
```diff
@@ -81,8 +81,10 @@ defmodule WebSock do
   * `{:stop, reason :: term(), close_detail(), state()}`: Similar to the previous clause, but allows
     for the explicit setting of either a plain close code or a close code with a body to be sent to
     the client
+  * `{:stop, reason :: term(), close_detail(), state()}`: Similar to the previous clause, but allows
+    for the sending of one or more frames before sending the connection close frame to the client
   """
-  @callback handle_in({message(), opcode: data_opcode()}, state()) :: handle_result()
+  @callback handle_in({payload(), opcode: data_opcode()}, state()) :: handle_result()
```
Comment (icehaunter) [inline lib/websock.ex]: Just curious, can iodata actually occur here in a callback, or is it always a binary?

**Review by icehaunter — APPROVED** [review]
Awesome! I think the documentation and spec is much clearer with your changes, and the feature itself fits our requirements perfectly well. Thank you very much.

**Comment by mtrudel** [comment]
Great! I'll let this stand for comments for a day or so and then get the other PRs ready to actually implement this

**Review by josevalim — APPROVED** [review]

File: lib/websock.ex (hunk 3 above)
Comment (mtrudel) [reply lib/websock.ex]: That's a good point! Cowboy [specs it](https://github.com/ninenines/cowboy/blob/master/src/cowboy_websocket.erl#L60) as a binary, and Bandit [explicitly flattens](https://github.com/mtrudel/bandit/blob/b1f6d9889dc38ecea790e14929f895972e5b2935/lib/bandit/websocket/connection.ex#L96) its internal iodata to match this API (recall that WebSock was basically built-to-fit the existing Cowboy API in order to facilitate Phoenix adoption, so a bunch of things don't make as much sense when cut from whole cloth). 

Note that `message` as used in callback returns may still be iodata (that is, WebSock will only ever pass you a binary, but will accept iodata in a `{:push,...}` or similar tuple).

I'll update the specs to reflect that.

Comment (mtrudel) [reply lib/websock.ex]: Resolved.

Comment (icehaunter) [reply lib/websock.ex]: Yeah, I gathered that you can push iodata, was curious explicitly about the callback argument. Good to know that I don't need to flatten it and that the spec reflects that now.

I think it makes sense that the callback receives a binary and not an iolist, makes it much easier to work with

File: lib/websock.ex [hunk 4]
```diff
@@ -81,8 +78,10 @@ defmodule WebSock do
   * `{:stop, reason :: term(), close_detail(), state()}`: Similar to the previous clause, but allows
     for the explicit setting of either a plain close code or a close code with a body to be sent to
     the client
+  * `{:stop, reason :: term(), close_detail(), state()}`: Similar to the previous clause, but allows
```
Comment (icehaunter) [inline lib/websock.ex]: Just noticed that this doc item references incorrect spec part (4-tuple, not the 5-tuple with messages).

```suggestion
  * `{:stop, reason :: term(), close_detail(), messages(), state()}`: Similar to the previous clause, but allows
```

Comment (mtrudel) [reply lib/websock.ex]: yeesh. I'm just full of 'em today.

Comment (mtrudel) [reply lib/websock.ex]: Fixed.

Comment (icehaunter) [reply lib/websock.ex]: I've reread the PR and I think it looks great, no further comments :)

**Comment by mtrudel** [comment]
Implementations for Bandit and Cowboy (the latter via websock_adapter) are linked above; I'll give those a day or so to soak up reviews (feel free to add your own!) and then I'll get all this merged

**Pull request merged**